test/subscription_test.dart

```diff
@@ -595,4 +595,145 @@ Future<void> main([List<String>? args]) async {
+
+    final unnamedResults = await query.subscribe(); // +1 unnamed subscription
+    await query.subscribe(); // +0 subscription already exists
+    await realm.all<Event>().subscribe(); // +0 subscription already exists
+    await queryFiltered.subscribe(); // +1 unnamed subscription
+    final namedResults1 = await query.subscribe(name: subscriptionName1); // +1 named subscription
+    final namedResults2 = await query.subscribe(name: subscriptionName2); // +1 named subscription
+    expect(realm.subscriptions.length, 4);
+
+    expect(query.unsubscribe(), isFalse); // -0 (query is not a subscription)
+    expect(realm.subscriptions.length, 4);
+
+    expect(unnamedResults.unsubscribe(), isTrue); // -1 unnamed subscription on query
+    expect(realm.subscriptions.length, 3);
+    expect(realm.subscriptions.find(queryFiltered), isNotNull);
+    expect(realm.subscriptions.findByName(subscriptionName1), isNotNull);
+    expect(realm.subscriptions.findByName(subscriptionName2), isNotNull);
+
+    realm.subscriptions.update((mutableSubscriptions) => mutableSubscriptions.removeAllUnnamed()); // -1 unnamed subscription on queryFiltered
+
+    expect(realm.subscriptions.length, 2);
+    expect(realm.subscriptions.findByName(subscriptionName1), isNotNull);
+    expect(realm.subscriptions.findByName(subscriptionName2), isNotNull);
+
+    expect(namedResults1.unsubscribe(), isTrue); // -1 named subscription sub1
+    expect(realm.subscriptions.length, 1);
+    expect(realm.subscriptions.findByName(subscriptionName2), isNotNull);
+
+    expect(namedResults2.unsubscribe(), isTrue); // -1 named subscription
+    expect(realm.subscriptions.length, 0);
+  });
+
+  baasTest('Flexible sync subscribe/unsubscribe API wait for download', (configuration) async {
+    final prefix = generateRandomString(4);
+    final byTestRun = "name BEGINSWITH '$prefix'";
+    App app = App(configuration);
+    final user = await app.logIn(Credentials.anonymous(reuseCredentials: false));
+    final config = Configuration.flexibleSync(user, [Product.schema]);
+    final realm = getRealm(config);
+    await realm.query<Product>(byTestRun).subscribe();
+    realm.write(() {
+      for (var i = 0; i < 20; i++) {
+        realm.add(Product(ObjectId(), "${prefix}_${i + 1}"));
+      }
+    });
+    await realm.syncSession.waitForUpload();
+    realm.close();
+
+    final user1 = await app.logIn(Credentials.anonymous(reuseCredentials: false));
+    final config1 = Configuration.flexibleSync(user1, [Product.schema]);
+    final realm1 = getRealm(config1);
+    final query = realm1.query<Product>('$byTestRun AND name ENDSWITH \$0', ["1"]);
+    final results = await query.subscribe(waitForSyncMode: WaitForSyncMode.never);
+    expect(results.length, 0); // didn't wait for downloading because of WaitForSyncMode.never
+
+    final first = await query.subscribe(waitForSyncMode: WaitForSyncMode.always, timeout: Duration(milliseconds: 1));
```

Comment:
Feels brittle to assume that a short wait time will not allow the download to succeed.